jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java

```diff
@@ -570,6 +596,29 @@ else if (len == -1)
             if (parser != null)
                 parser.parse(BufferUtil.EMPTY_BUFFER, true);
         }
+        finally
+        {
+            boolean cleanup = false;
+            synchronized (this)
+            {
+                switch (state)
+                {
+                    case PARSING:
+                        state = State.PARSED;
+                        break;
+
+                    case CLOSING:
+                        cleanup = true;
+                        break;
+
+                    default:
+                        _err = new IllegalStateException(state.name());
+                }
+            }
+
+            if (cleanup)
```

Comment:
Save the boolean and do the cleanup in the appropriate case.

Reply:
I did it in this style because Simone always tells me to act outside of the synchronized block.

Reply:
Ah yes, I guess doing the delete will take some time and you don't want to hold the lock while you're doing it.

Regarding the deletes, I think it would be better if the state transition was handled in only 1 place - here its in the finally block of parse() and also in deleteParts() method.

Reply:
I had the state transition only in `deleteParts()` but that had the effect that if you called `deleteParts()` twice, it would not wait for the parsing thread and do the cleanup straight away. So I think it's better to do it here as well.

So now
The initial deleteParts will do `PARSING->ERROR`
Calling deleteParts again will be a noop from `ERROR` or `DELETED` state.
The parsing thread does `ERROR->DELETED` and does the `uncheckedDeleteParts()`

Reply:
I still think it would be a cleaner solution if the deletion happened only in 1 place, and that is in deleteParts(). If deleteParts() gets called and the parser is still parsing, it should set the ERROR state, wait for the parser to stop, and then do a full clean. Is there something I'm missing that makes that a bad idea?

Also, it would be a good idea to add a description of each of those states in the javadoc state transition diagram: especially so it's clear that ERROR state only happens one way, and that is that deleteParts() is called by another thread when parse() is happening in a different thread, otherwise it could easily be misinterpreted that ERROR state represents a parsing error.